File: app/src/main/java/james/metronome/services/MetronomeService.java
package james.metronome.services;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.SoundPool;
import android.os.Binder;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import james.metronome.R;
import james.metronome.views.TicksView;

public class MetronomeService extends Service implements Runnable {

    public static final String ACTION_START = "james.metronome.ACTION_START";
    public static final String ACTION_PAUSE = "james.metronome.ACTION_PAUSE";
    public static final String EXTRA_BPM = "james.metronome.EXTRA_BPM";

    public static final String PREF_TICK = "tick";
    public static final String PREF_INTERVAL = "interval";
    public static final String PREF_EMPHASIS_SIZE = "emphasisSize";
    public static final String PREF_EMPHASIS = "emphasis";

    private final IBinder binder = new LocalBinder();

    private SharedPreferences prefs;
    private int bpm;
    private long interval;

    private SoundPool soundPool;
    private Handler handler;
    private int soundId = -1;
    private boolean isPlaying;

    private List<Boolean> emphasisList = new ArrayList<>(Arrays.asList(new Boolean[]{true, true, true, true}));
    private int emphasisIndex;

    private Vibrator vibrator;

    private TickListener listener;

    @Override
    public void onCreate() {
        super.onCreate();
        prefs = PreferenceManager.getDefaultSharedPreferences(this);

        vibrator = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            soundPool = new SoundPool.Builder()
                    .setMaxStreams(1)
                    .setAudioAttributes(new AudioAttributes.Builder()
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .build())
                    .build();
        } else soundPool = new SoundPool(1, AudioManager.STREAM_MUSIC, 0);

        int tick = prefs.getInt(PREF_TICK, 0);
        if (!TicksView.ticks[tick].isVibration())
            soundId = TicksView.ticks[tick].getSoundId(this, soundPool);

        interval = prefs.getLong(PREF_INTERVAL, 500);
        bpm = toBpm(interval);

        emphasisList = new ArrayList<>();
        int emphasisSize = prefs.getInt(PREF_EMPHASIS_SIZE, 4);
        for (int i = 0; i < emphasisSize; i++) {
            emphasisList.add(prefs.getBoolean(PREF_EMPHASIS + i, false));
        }

        handler = new Handler();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null && intent.getAction() != null) {
            switch (intent.getAction()) {
                case ACTION_START:
                    setBpm(intent.getIntExtra(EXTRA_BPM, bpm));
                    pause();
                    play();
                    break;
                case ACTION_PAUSE:
                    pause();
                    break;
            }
        }
        return START_STICKY;
    }

    private static int toBpm(long interval) {
        return (int) (60000 / interval);
    }

    private static long toInterval(int bpm) {
        return (long) 60000 / bpm;
    }

    public void play() {
        handler.post(this);
        isPlaying = true;
        emphasisIndex = 0;

        Intent intent = new Intent(this, MetronomeService.class);
        intent.setAction(ACTION_PAUSE);

        NotificationCompat.Builder builder;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            ((NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE))
                    .createNotificationChannel(new NotificationChannel("metronome", getString(R.string.app_name), NotificationManager.IMPORTANCE_DEFAULT));

            builder = new NotificationCompat.Builder(this, "metronome");
        } else
            builder = new NotificationCompat.Builder(this);

        startForeground(530,
                builder.setContentTitle(getString(R.string.notification_title))
                        .setContentText(getString(R.string.notification_desc))
                        .setSmallIcon(R.drawable.ic_notification)
                        .setContentIntent(PendingIntent.getService(this, 0, intent, PendingIntent.FLAG_ONE_SHOT))
                        .setPriority(NotificationCompat.PRIORITY_LOW)
                        .build()
        );

        if (listener != null)
            listener.onStartTicks();
    }

    public void pause() {
        handler.removeCallbacks(this);
        stopForeground(true);
        isPlaying = false;

        if (listener != null)
            listener.onStopTicks();
    }

    public void setEmphasisList(List<Boolean> emphasisList) {
        this.emphasisList = emphasisList;
        emphasisIndex = 0;

        SharedPreferences.Editor editor = prefs.edit();
        editor.putInt(PREF_EMPHASIS_SIZE, emphasisList.size());
        for (int i = 0; i < emphasisList.size(); i++) {
            editor.putBoolean(PREF_EMPHASIS + i, emphasisList.get(i));
        }
        editor.apply();
    }

    public void setBpm(int bpm) {
        this.bpm = bpm;
        interval = toInterval(bpm);
        prefs.edit().putLong(PREF_INTERVAL, interval).apply();
        if (listener != null)
            listener.onBpmChanged(bpm);
    }

    public void setTick(int tick) {
        if (!TicksView.ticks[tick].isVibration()) {
            soundId = TicksView.ticks[tick].getSoundId(this, soundPool);
            if (!isPlaying)
                soundPool.play(soundId, 1.0f, 1.0f, 0, 0, 1.0f);
        } else soundId = -1;

        prefs.edit().putInt(PREF_TICK, tick).apply();
    }

    public boolean isPlaying() {
        return isPlaying;
    }

    public long getInterval() {
        return interval;
    }

    public List<Boolean> getEmphasisList() {
        return emphasisList;
    }

    public int getBpm() {
        return bpm;
    }

    public int getTick() {
        return prefs.getInt(PREF_TICK, 0);
    }

    public void setTickListener(TickListener listener) {
        this.listener = listener;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public boolean onUnbind(Intent intent) {
        listener = null;
        return super.onUnbind(intent);
    }

    @Override
    public void onDestroy() {
        handler.removeCallbacks(this);
        super.onDestroy();
    }

    @Override
    public void run() {
        if (isPlaying) {
            handler.postDelayed(this, interval);

            if (emphasisIndex >= emphasisList.size())
                emphasisIndex = 0;
            boolean isEmphasis = emphasisList.get(emphasisIndex);
            if (listener != null)
                listener.onTick(isEmphasis, emphasisIndex);
            emphasisIndex++;

            if (soundId != -1)
                soundPool.play(soundId, 1, 1, 0, 0, isEmphasis ? 1.5f : 1);
            else if (Build.VERSION.SDK_INT >= 26)
                vibrator.vibrate(VibrationEffect.createOneShot(isEmphasis ? 50 : 20, VibrationEffect.DEFAULT_AMPLITUDE));
            else vibrator.vibrate(isEmphasis ? 50 : 20);
        }
    }

    public class LocalBinder extends Binder {
        public MetronomeService getService() {
            return MetronomeService.this;
        }
    }

    public interface TickListener {
        void onStartTicks();

        void onTick(boolean isEmphasis, int index);

        void onBpmChanged(int bpm);

        void onStopTicks();
    }
}

File: app/src/main/java/james/metronome/services/MetronomeService.kt
package james.metronome.services

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.SoundPool
import android.os.*
import android.preference.PreferenceManager
import androidx.core.app.NotificationCompat
import james.metronome.R
import james.metronome.views.TicksView
import java.util.*

class MetronomeService : Service(), Runnable {

    private val binder: IBinder = LocalBinder()
    private val prefs: SharedPreferences by lazy { PreferenceManager.getDefaultSharedPreferences(this) }

    var interval: Long
        get() = prefs.getLong(PREF_INTERVAL, 500)
        set(value) {
            prefs.edit().putLong(PREF_INTERVAL, value).apply()
            listener?.onBpmChanged(toBpm(value))
        }

    var bpm: Int
        get() = toBpm(interval)
        set(value) {
            interval = toInterval(value)
        }

    private val soundPool: SoundPool by lazy {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            SoundPool.Builder()
                    .setMaxStreams(1)
                    .setAudioAttributes(AudioAttributes.Builder()
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .build())
                    .build()
        } else SoundPool(1, AudioManager.STREAM_MUSIC, 0)
    }

    private val handler: Handler = Handler()

    var tick: Int
        get() = prefs.getInt(PREF_TICK, 0)
        set(tick) {
            prefs.edit().putInt(PREF_TICK, tick).apply()
        }

    private val soundId: Int
        get() {
            val tickIndex = tick
            return if (!TicksView.ticks[tickIndex].isVibration)
                TicksView.ticks[tickIndex].getSoundId(this, soundPool)
            else -1
        }

    var isPlaying = false
        private set

    var emphasisList: MutableList<Boolean>
        get() = ArrayList<Boolean>().apply {
            val emphasisSize = prefs.getInt(PREF_EMPHASIS_SIZE, 4)
            for (i in 0 until emphasisSize) {
                add(prefs.getBoolean(PREF_EMPHASIS + i, false))
            }
        }
        set(value) {
            emphasisIndex = 0
            prefs.edit().apply {
                putInt(PREF_EMPHASIS_SIZE, value.size)
                for (i in value.indices) {
                    putBoolean(PREF_EMPHASIS + i, value[i])
                }

                apply()
            }
        }

    private var emphasisIndex = 0
    private val vibratorService: Vibrator by lazy { getSystemService(Context.VIBRATOR_SERVICE) as Vibrator }
    private val notificationManager: NotificationManager by lazy { getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager }
    private var listener: TickListener? = null

    override fun onCreate() {
        super.onCreate()
        interval = prefs.getLong(PREF_INTERVAL, 500)
        bpm = toBpm(interval)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_START -> {
                bpm = intent.getIntExtra(EXTRA_BPM, bpm)
                pause()
                play()
            }
            ACTION_PAUSE -> pause()
        }

        return START_STICKY
    }

    fun play() {
        handler.post(this)
        isPlaying = true
        emphasisIndex = 0
        val intent = Intent(this, MetronomeService::class.java).apply { action = ACTION_PAUSE }

        val builder: NotificationCompat.Builder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notificationManager.createNotificationChannel(NotificationChannel("metronome", getString(R.string.app_name), NotificationManager.IMPORTANCE_LOW))
            NotificationCompat.Builder(this, "metronome")
        } else NotificationCompat.Builder(this)

        startForeground(530,
                builder.setContentTitle(getString(R.string.notification_title))
                        .setContentText(getString(R.string.notification_desc))
                        .setSmallIcon(R.drawable.ic_notification)
                        .setContentIntent(PendingIntent.getService(this, 0, intent, PendingIntent.FLAG_ONE_SHOT))
                        .setPriority(NotificationCompat.PRIORITY_LOW)
                        .build()
        )

        listener?.onStartTicks()
    }

    fun pause() {
        handler.removeCallbacks(this)
        stopForeground(true)

        isPlaying = false
        listener?.onStopTicks()
    }

    fun setTickListener(listener: TickListener?) {
        this.listener = listener
    }

    override fun onBind(intent: Intent): IBinder? {
        return binder
    }

    override fun onUnbind(intent: Intent): Boolean {
        listener = null
        return super.onUnbind(intent)
    }

    override fun onDestroy() {
        handler.removeCallbacks(this)
        super.onDestroy()
    }

    override fun run() {
        if (isPlaying) {
            handler.postDelayed(this, interval)
            if (emphasisIndex >= emphasisList.size) emphasisIndex = 0
            val isEmphasis = emphasisList[emphasisIndex]
            listener?.onTick(isEmphasis, emphasisIndex)
            emphasisIndex++

            when {
                soundId != -1 -> soundPool.play(soundId, 1f, 1f, 0, 0, if (isEmphasis) 1.5f else 1f)
                Build.VERSION.SDK_INT >= 26 -> vibratorService.vibrate(VibrationEffect.createOneShot(if (isEmphasis) 50 else 20.toLong(), VibrationEffect.DEFAULT_AMPLITUDE))
                else -> vibratorService.vibrate(if (isEmphasis) 50 else 20.toLong())
            }
        }
    }

    inner class LocalBinder : Binder() {
        val service: MetronomeService
            get() = this@MetronomeService
    }

    interface TickListener {
        fun onStartTicks()
        fun onTick(isEmphasis: Boolean, index: Int)
        fun onBpmChanged(bpm: Int)
        fun onStopTicks()
    }

    companion object {
        const val ACTION_START = "james.metronome.ACTION_START"
        const val ACTION_PAUSE = "james.metronome.ACTION_PAUSE"
        const val EXTRA_BPM = "james.metronome.EXTRA_BPM"
        const val PREF_TICK = "tick"
        const val PREF_INTERVAL = "interval"
        const val PREF_EMPHASIS_SIZE = "emphasisSize"
        const val PREF_EMPHASIS = "emphasis"

        private fun toBpm(interval: Long): Int {
            return (60000 / interval).toInt()
        }

        private fun toInterval(bpm: Int): Long {
            return 60000.toLong() / bpm
        }
    }
}